### ISSUE TITLE
Enforce stricter slice operations

### ISSUE THREAD

**Issue by orsinium** [wemake-services/wemake-python-styleguide#1011]
# Rule request

## Thesis

The list has a set of methods that duplicates operations that you can do with `+` operator or slices:

```python
# bad
items += [item]
# good
items.append(item)

# bad
items = [item] + items
# good
items.insert(0, item)

# bad
items += other
# good
items.extend(other)

# bad
items = items[::-1]
# good
items.reverse()

# bad
items = sorted(items)
# good
items.sort()

# bad
other = items[:]
# good
other = items.copy()

# bad
items[:] = []
# good
items.clear()

# bad
items = items[:-1]
# good
items.pop()
```

## Reasoning

We have two ways to do something. Let's prefer a more descriptive way.

This issue contains a lot of related cases. Consider splitting it by smaller ones.

**Comment by orsinium** [comment]
If you agree and it's implementable, I can later do the same issue for sets, that's also fun.

**Comment by sobolevn** [comment]
Let's implement several rules that are not bound to types:

- `items[::-1]` should not be used. Use `reversed` or `.reverse()`
- `items[:]` should not be used. Use `.copy()`
- `items = items[:-1]` should not be used. Use `items.pop()`

@orsinium as always - awesome suggestions! Thanks!

**Comment by Eric4Jiang** [comment]
Hi! I would love to contribute and try to resolve this issue if no one else is on it.

**Comment by sobolevn** [comment]
Sure, @Eric4Jiang! Thanks a lot! Assigned.

**Comment by Eric4Jiang** [comment]
@sobolevn 

Hi! Would ```items[::]``` fall into the same category as ```items[:]``` because it can be replaced with ```.copy()```?

**Comment by sobolevn** [comment]
Sure!

**Comment by orsinium** [comment]
Hmmm, that gives me an idea for one more rule. I've made #1071. Thank you, @Eric4Jiang!

**Comment by bDrwx** [comment]
Hi, @sobolevn. I'm gonna try to make this issue. Is it still relevant?

**Comment by orsinium** [comment]
@bDrwx Thank you! Looks like there is still no implementation for the rule. Go ahead! I've assigned you to the issue 👍🏿

**Comment by Mash378** [comment]
Has the issue been resolved? If not I will gladly work to solve it

**Comment by skv0zsneg** [comment]
Hi @sobolevn and @orsinium!

Not sure but it's been 50 days since this issue was assigned to @Mash378 and as I can see did not moved. If no nobody minds I can take it 🙃

**Comment by sobolevn** [comment]
Thanks! ❤

**Comment by skv0zsneg** [comment]
@sobolevn where is the best place for this rule can be? I think consistency is a better choice but maybe in best practice?

**Comment by skv0zsneg** [comment]
After some analyse I think it can be solve this way

This is the violations of non strict operation with slice: 

```python
# bad
items = items[::-1]
# good
items.reverse()

# bad
other = items[:]
# good
other = items.copy()

# bad
items = items[:-1]
# good
items.pop()
```

This is the violations of non strict operation with list:

```python
# bad
items += other
# good
items.extend(other)

# bad
items += [item]
# good
items.append(item)

# bad
items = [item] + items
# good
items.insert(0, item)

# bad
items[:] = []
# good
items.clear()
```

This is the violations of non strict operation with iterable:

```python
# bad
items = sorted(items)
# good
items.sort()
```

So, every violations above must have own `ASTViolation` subclasses. And maybe 'Best Practice' a better type for them 🙃

**Comment by sobolevn** [comment]
Let's only cover

```python
# bad
items = items[::-1]
# good
items.reverse()

# bad
other = items[:]
# good
other = items.copy()

# bad
items = items[:-1]
# good
items.pop()
```

for now :)

**Comment by y-polyakov-ddl** [comment]
Would there be a one-line `#good` way to do this (currently-`#bad`) operation?
```py3
# bad
items[:] = other
# good?
items.clear()
items.extend(other)
```
